kong/resty/dns_client/init.lua

```diff
@@ -0,0 +1,782 @@
+local timer_at  = ngx.timer.at
+
+local type          = type
+local pairs         = pairs
+local ipairs        = ipairs
+local math_min      = math.min
+local string_lower  = string.lower
+local table_insert  = table.insert
+
+local parse_hosts   = utils.parse_hosts
+local ipv6_bracket  = utils.ipv6_bracket
+local search_names  = utils.search_names
+local get_next_round_robin_answers     = utils.get_next_round_robin_answers
+local get_weighted_round_robin_answers = utils.get_weighted_round_robin_answers
+
+local req_dyn_hook_run_hooks = require("kong.dynamic_hook").run_hooks
+local cycle_aware_deep_copy  = require("kong.tools.utils").cycle_aware_deep_copy
+
+
+-- Constants and default values
+
+local DEFAULT_ERROR_TTL   = 1     -- unit: second
+local DEFAULT_STALE_TTL   = 4
+local DEFAULT_EMPTY_TTL   = 30
+-- long-lasting TTL of 10 years for hosts or static IP addresses in cache settings
+local LONG_LASTING_TTL    = 10 * 365 * 24 * 60 * 60
+
+local DEFAULT_ORDER = { "LAST", "SRV", "A", "AAAA", "CNAME" }
+
+local TYPE_SRV      = resolver.TYPE_SRV
+local TYPE_A        = resolver.TYPE_A
+local TYPE_AAAA     = resolver.TYPE_AAAA
+local TYPE_CNAME    = resolver.TYPE_CNAME
+local TYPE_LAST     = -1
+
+local NAME_TO_TYPE = {
+  SRV     = TYPE_SRV,
+  A       = TYPE_A,
+  AAAA    = TYPE_AAAA,
+  CNAME   = TYPE_CNAME,
+  LAST    = TYPE_LAST,
+}
+
+local TYPE_TO_NAME = {
+  [TYPE_SRV]      = "SRV",
+  [TYPE_A]        = "A",
+  [TYPE_AAAA]     = "AAAA",
+  [TYPE_CNAME]    = "CNAME",
+}
+
+local HIT_L3 = 3 -- L1 lru, L2 shm, L3 callback, L4 stale
+
+local HIT_LEVEL_TO_NAME = {
+  [1] = "hit_lru",
+  [2] = "hit_shm",
+  [3] = "hit_cb",
+  [4] = "hit_stale",
+}
+
+-- server replied error from the DNS protocol
+local NAME_ERROR_CODE             = 3 -- response code 3 as "Name Error" or "NXDOMAIN"
+-- client specific error
+local CACHE_ONLY_ERROR_CODE       = 100
+local CACHE_ONLY_ERROR_MESSAGE    = "cache only lookup failed"
+local CACHE_ONLY_ANSWERS = tablex.readonly({ errcode = CACHE_ONLY_ERROR_CODE, errstr = CACHE_ONLY_ERROR_MESSAGE })
+local EMPTY_RECORD_ERROR_CODE     = 101
+local EMPTY_RECORD_ERROR_MESSAGE  = "empty record received"
+
+
+-- APIs
+
+local _M = {}
+local mt = { __index = _M }
+
+-- copy TYPE_*
+for k,v in pairs(resolver) do
+  if type(k) == "string" and k:sub(1,5) == "TYPE_" then
+    _M[k] = v
+  end
+end
+_M.TYPE_LAST = TYPE_LAST
+
+
+local TRIES_MT = { __tostring = cjson.encode, }
+
+
+local function stats_init(stats, name)
+  if not stats[name] then
+    stats[name] = {}
+  end
+end
+
+
+local function stats_count(stats, name, key)
+  stats[name][key] = (stats[name][key] or 0) + 1
+end
+
+
+local function stats_set(stats, name, key, value)
+  stats[name][key] = value
+end
+
+
+-- lookup or set TYPE_LAST (the DNS record type from the last successful query)
+local function insert_last_type(cache, name, qtype)
+  local key = "last:" .. name
+  if TYPE_TO_NAME[qtype] and cache:get(key) ~= qtype then
+    cache:set(key, { ttl = 0 }, qtype)
```

Comment:
I think that `{ ttl = 0 }` could be defined as a module level constant to avoid creating temporary table.